sql/core/src/main/scala/org/apache/spark/sql/classic/Dataset.scala

```diff
@@ -1544,6 +1544,35 @@ class Dataset[T] private[sql](
     }
   }
 
+  /**
+   * Repartitions the Dataset using the specified partition ID expression.
+   *
+   * The number of partitions will be configured by `spark.sql.shuffle.partitions`.
+   *
+   * @param partitionIdExpr the expression to be used as the partition ID. Must be an integral type.
+   *
+   * @group typedrel
+   * @since 4.1.0
+   */
+  def repartitionById(partitionIdExpr: Column): Dataset[T] = {
+    repartitionById(sparkSession.sessionState.conf.numShufflePartitions, partitionIdExpr)
+  }
+
+  /**
+   * Repartitions the Dataset into the given number of partitions using the specified
+   * partition ID expression.
+   *
+   * @param numPartitions the number of partitions to use.
+   * @param partitionIdExpr the expression to be used as the partition ID. Must be an integral type.
+   *
+   * @group typedrel
+   * @since 4.1.0
+   */
+  def repartitionById(numPartitions: Int, partitionIdExpr: Column): Dataset[T] = {
+    val directShufflePartitionIdCol = Column(DirectShufflePartitionID(partitionIdExpr.expr))
+    repartitionByExpression(Some(numPartitions), Seq(directShufflePartitionIdCol))
```

Comment:
We can create `RepartitionByExpression` directly with a special flag to indicate pass through, then we don't need `DirectShufflePartitionID`.